File: leetcode-go/Q001-Q050/Q045_Jump_Game_II/Q045_Jump_Game_II.go
package Q045_Jump_Game_II

func jump(nums []int) int {
	maxpos := 0
	curpos := 0
	step := 0
	for i := 0; i < len(nums)-1; i++ {
		maxpos = max(maxpos, nums[i]+i)
		if curpos == i {
			curpos = maxpos
			step++
		}
	}
	return step
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: src/Y2018/facebook/Q045_Jump_Game_II.java
package Y2018.facebook;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

public class Q045_Jump_Game_II {

    // BFS: Time Limit Exceeded
    public int jump(int[] nums) {
        Queue<Integer> Q = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        Q.offer(0);
        visited.add(0);
        int result = 0;
        while (!Q.isEmpty()) {
            int size = Q.size();
            while (size-- > 0) {
                int pos = Q.poll();
                if (pos == nums.length - 1) {
                    return result;
                }
                for (int i = 1; i <= nums[pos] && pos + i < nums.length; i++) {
                    if (!visited.contains(pos + i)) {
                        Q.offer(pos + i);
                        visited.add(pos + i);
                    }
                }
            }
            result++;
        }
        return result;
    }

}

